### PR TITLE
4167 enhance bundle `run` logic

### LINKED ISSUE #4167: Need to enhance the logic control of bundle config
**Is your feature request related to a problem? Please describe.**
There are 2 missing features of the logic control in bundle `run` script:
1. Currently, we have key `_requires_` to execute some logic before instantiation of the component, but if the component has other dependent components as args, we can't guarantee to run before the instantiation of other components.
2. `_requires_` can only run before the component, can't execute some code after the component.
For example, can't run `ddp_init` before `network` instantiation, can't run `destroy_ddp_group()` after `trainer.run()`:
```json
"trainer": {
    "_target_": "SupervisedTrainer",
    "_requires_": ["@ddp_init", "@determinism", "@cudnn_opt"],
    "max_epochs": 100,
    "device": "@device",
    "train_data_loader": "@train#dataloader",
    "network": "@network",
    "loss_function": "@loss",
    "optimizer": "@optimizer",
    "inferer": "@train#inferer",
    "postprocessing": "@train#postprocessing",
    "key_train_metric": "@train#key_metric",
    "train_handlers": "@train#handlers",
    "amp": true
}
```

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #4167 .

### Description
This PR enhanced the execution logic of bundler `run`.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @wyli ,

I updated the PR according to your comments, could you please help review it again?

Thanks in advance.

**Review by ericspod — APPROVED** [review]
Looks good to me, this is more general.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Thanks for your review, a typical multi-gpu usage can be:
```json
"training": [
    "$dist.init_process_group(backend='nccl')",
    "$torch.cuda.set_device(@device)",
    "$monai.utils.set_determinism(seed=123)",
    "$@train#trainer.run()",
    "$dist.destroy_process_group()"
]
```
Currently, we use `Set()` to save references which can't guarantee the list order:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/reference_resolver.py#L250
Let me try to also update it in this PR.

Thanks.

**Comment by ericspod** [comment]
> Currently, we use python `Set` to save references which can't guarantee the list order: https://github.com/Project-MONAI/MONAI/blob/dev/monai/bundle/reference_resolver.py#L250 Let me try to also update it in this PR.
> 
> Thanks.

Yes we should use list also because statements might be duplicated.

File: monai/bundle/reference_resolver.py [hunk 1]
```diff
@@ -247,15 +247,17 @@ def find_refs_in_config(cls, config, id: str, refs: Optional[Set[str]] = None) -
             refs: list of the ID name of found references, default to `None`.
 
         """
-        refs_: Set[str] = refs or set()
+        refs_: List[str] = refs or []
         if isinstance(config, str):
-            return refs_.union(cls.match_refs_pattern(value=config))
+            for i in cls.match_refs_pattern(value=config):
+                if i not in refs_:
```
Comment (wyli) [inline monai/bundle/reference_resolver.py:253]: It'll be nice to use a dictionary or a similar data structure to ensure unique references, better readability, and slightly better perf for the `in` checks.

Comment (Nic-Ma) [reply monai/bundle/reference_resolver.py]: Sure, updated in the latest commit.
Could you please help review it again?

Thanks in advance.

**Pull request merged**